### PR TITLE
Use SDK Registry to distribute 1.0 snapshots

### PR DESCRIPTION
**Pull request by LukasPaczos**
This PR adds the configuration to publish Core and UI 1.0 snapshots to SDK Registry and enables the snapshot generation on every push to `master`. Currently, we can host only one version of a snapshot (for example `1.0.0-SNAPSHOT`), so the artifacts will keep overriding each other until the timestamped snapshot feature is implemented on the SDK Registry side. If you need to have a snapshot that's not going to get overwritten, hardcode the `version` field in the publish script with some custom name. We can introduce an automated way of doing so in the future.

The snapshots repo configuration is:
```
maven {
    url 'https://api.mapbox.com/downloads/v2/snapshots/maven'
    authentication {
        basic(BasicAuthentication)
    }
    credentials {
        username = "mapbox"
        password = "{secret_token}"
    }
}
```

I also had to update the reference app's crashlytics to bump the Android Gradle Plugin.

The `scripts/publish_release_to_aws.sh` and the `sdkRegistryPublish` Gradle tasks could be something that we could migrate to a mapbox plugin that you've been working on @pengdev. After this change, we still need the Bintray configuration to resolve the Nav Native artifact downstream but if schedules permit I'd like to wait with that implementation until we have a centralized plugin for all projects instead of duplicating the code.

### CODE REVIEW

**Comment by pengdev** [comment]
This is looking 👍 , I'm wondering if we could package the `publish_release_to_aws.sh` shell script with the sdk-registry gradle plugin. 

Also I'm thinking about the aws client, shall we assume the aws client available in the environment with all the credentials validated? There is also an option to use java aws sdk in the gradle plugin and have the options exposed for users to provide the AWS credentials. 

@LukasPaczos do you have suggestions of above questions? Anyway I think using the command line tool is an easy way to get started.

File: libnavigation-ui/build.gradle [hunk 1]
```diff
@@ -3,7 +3,6 @@ apply plugin: 'kotlin-android'
 apply plugin: 'org.jetbrains.dokka'
 apply plugin: 'kotlin-android-extensions'
 apply plugin: 'com.jaredsburrows.license'
-apply plugin: 'com.mapbox.android.sdk.versions'
```
Comment (Guardiola31337) [inline libnavigation-ui/build.gradle:6]: Is it safe to remove https://github.com/mapbox/android-sdk-versions-plugin 🤔? cc @harvsu

File: libnavigation-core/build.gradle [hunk 2]
```diff
@@ -3,7 +3,6 @@ apply plugin: 'kotlin-android'
 apply plugin: 'kotlin-kapt'
 apply plugin: 'org.jetbrains.dokka'
 apply plugin: 'com.jaredsburrows.license'
-apply plugin: 'com.mapbox.android.sdk.versions'
```
Comment (Guardiola31337) [inline libnavigation-core/build.gradle:6]: Same as above re: https://github.com/mapbox/android-sdk-versions-plugin

File: libnavigation-ui/build.gradle (hunk 1 above)
Comment (LukasPaczos) [reply libnavigation-ui/build.gradle:6]: Good catch! The semantic versioning has been blocking me when using the plugin because it requires the global variables to be set which impacted the scripts. I can try re-adding the plugin and extending the scripts to make it work again.

Comment (harvsu) [reply libnavigation-ui/build.gradle:6]: Removing this plugin would affect the reformed user Agent in Events okhttp request. This generator plugin generates a file that feeds the Nav SDK version to the reformed user agent. c.c @alfwatt

File: gradle/dependencies.gradle [hunk 3]
```diff
@@ -50,7 +50,7 @@ ext {
       kotlinStdLib              : '1.3.72',
       ankoCommon                : '0.10.0',
       firebaseCore              : '16.0.7',
-      crashlytics               : '2.9.9',
+      crashlytics               : '17.1.0',
```
Comment (RingerJK) [inline gradle/dependencies.gradle:53]: should it be renamed according firebase names, just like with `firebaseCore `
```suggestion
      firebaseCrashlytics       : '17.1.0',
```

btw https://firebase.google.com/support/release-notes/android#kotlin+ktx
> The Firebase Android library firebase-core is no longer needed. This SDK included the Firebase SDK for Google Analytics.

File: gradle/dependencies.gradle [hunk 4]
```diff
@@ -165,17 +165,16 @@ ext {
...
       dependencyGraph  : '0.5.0',
       dependencyUpdates: '0.20.0',
       kotlin           : '1.3.72',
       license          : '0.8.5',
       jacoco           : '0.1.4',
       playPublisher    : '2.3.0',
-      googleServices   : '4.2.0',
-      crashlytics      : '1.26.1',
+      googleServices   : '4.3.3',
+      crashlytics      : '2.2.0',
```
Comment (RingerJK) [inline gradle/dependencies.gradle:176]:
```suggestion
      firebaseCrashlytics      : '2.2.0',
```

File: gradle/publish.gradle [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+
+                artifact(androidSourcesJar)
+                artifact(androidJavadocsJar)
+            }
+        }
+    }
+}
+
+bintray {
+    user = mapboxBintrayUser
```
Comment (RingerJK) [inline gradle/publish.gradle:53]: user and key mustn't by null or empty so may be add check for it and trows obvious exception.

File: scripts/publish_release_to_aws.sh [hunk 6]
```diff
@@ -0,0 +1,75 @@
...
+variant="$4"
+if [ -z "$module" ]; then
+  echo >&2 "First argument (module name) is missing."
+  exit 1
+fi
+if [ -z "$artifact_id" ]; then
+  echo >&2 "Second argument (Maven artifact ID) is missing."
+  exit 1
+fi
+if [ -z "$version" ]; then
```
Comment (RingerJK) [inline scripts/publish_release_to_aws.sh:42]: should we have a regex to check version?

File: gradle/publish.gradle (hunk 5 above)
Comment (RingerJK) [reply gradle/publish.gradle:53]: better place is `artifact-settings.gradle` file, wdyt?

Comment (LukasPaczos) [reply gradle/publish.gradle:53]: This variable is coming from `artifact-settings.gradle`

File: scripts/publish_release_to_aws.sh (hunk 6 above)
Comment (LukasPaczos) [reply scripts/publish_release_to_aws.sh:42]: We shouldn't enforce it on this step. There's the telem plugin mentioned above that does check it already. Here, we could hardcode a snapshot version that doesn't follow semver.

File: gradle/dependencies.gradle (hunk 3 above)
Comment (LukasPaczos) [reply gradle/dependencies.gradle:53]: This is crashlytics, not core.

Comment (LukasPaczos) [reply gradle/dependencies.gradle]: Ah, I think I got it, removed the dep :D

**Comment by LukasPaczos** [comment]
Ready for another round.

File: gradle/dependencies.gradle (hunk 4 above)
Comment (RingerJK) [reply gradle/dependencies.gradle:174]: seems version of plugin wasn't changed

**Comment by RingerJK** [comment]
regarding comment https://github.com/mapbox/mapbox-navigation-android/pull/3296/files#r452152135 suggestion: add to `artifact-settings.gradle`, check that Bintray User and Bintray Token is not null or empty so 

`artifact-settings.gradle`
```
    mapboxBintrayUser = project.hasProperty('BINTRAY_USER') ? project.property('BINTRAY_USER') : System.getenv('BINTRAY_USER')
    mapboxBintrayApiKey = project.hasProperty('BINTRAY_API_KEY') ? project.property('BINTRAY_API_KEY') : System.getenv('BINTRAY_API_KEY')
  if(mapboxBintrayUser.isNullOrEmpty() || mapboxBintrayApiKey.isNullOrEmpty()){
    trows new Exception("Non valid credentials: mapboxBintrayUser and mapboxBintrayApiKey mustn't be null or empty") 
  }
```

**Comment by LukasPaczos** [comment]
This PR has been refactored to use the SDK Registry plugin (thanks @pengdev!). Ready for another round of reviews.

**Comment by pengdev** [comment]
@LukasPaczos In the latest refactor, I used `aws s3 sync` instead of `aws s3 copy`, which requires additional permissions [refs](https://stackoverflow.com/questions/48894886/aws-permissions-required-for-sync)

**Comment by LukasPaczos** [comment]
Ah okay, thanks @pengdev. I blindly tried only `ListObjectsV2` but that wasn't enough. Let me update.

**Comment by LukasPaczos** [comment]
@pengdev I'm hitting
```
fatal error: An error occurred (AccessDenied) when calling the ListObjectsV2 operation: Access Denied
```

I understand that `ListObjectsV2` is explicitly required as well after all? Or should it be wrapped in any of the already included ones?

**Comment by LukasPaczos** [comment]
Hmm, capturing from [the docs](https://docs.aws.amazon.com/AmazonS3/latest/API/API_ListObjectsV2.html), it should be available with the new policy:
```
To use this operation in an AWS Identity and Access Management (IAM) policy, you must have permissions to perform the s3:ListBucket action.
```

Let me try again.

**Comment by LukasPaczos** [comment]
Updated the mapbox plugin to `v0.2.1` that changes the required policies and the upload mechanism, retrying now.

**Comment by LukasPaczos** [comment]
The artifacts are published correctly :tada: 

This is ready for another round of reviews.
